controller/src/state/world_state.rs

```diff
@@ -0,0 +1,172 @@
+use plane_core::{
+    messages::{
+        agent,
+        state::{
+            BackendMessageType, ClusterStateMessage, DroneMessageType, DroneMeta, WorldStateMessage,
+        },
+    },
+    types::{BackendId, ClusterName, DroneId},
+};
+use std::{
+    collections::{BTreeSet, HashMap, VecDeque},
+    net::IpAddr,
+    sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard},
+};
+
+#[derive(Default, Debug, Clone)]
+pub struct StateHandle(Arc<RwLock<WorldState>>);
+
+impl StateHandle {
+    pub fn new(world_state: WorldState) -> Self {
+        Self(Arc::new(RwLock::new(world_state)))
+    }
+
+    pub fn get_ready_drones(
+        &self,
+        cluster: &ClusterName,
+        timestamp: DateTime<Utc>,
+    ) -> Result<Vec<DroneId>> {
+        let world_state = self.0.write().expect("Could not acquire world_state lock.");
+
+        let min_keepalive = timestamp - chrono::Duration::seconds(30);
+
+        tracing::info!(?world_state, "World state");
+
+        Ok(world_state
+            .cluster(cluster)
+            .ok_or_else(|| anyhow!("Cluster not found."))?
+            .drones
+            .iter()
+            .filter(|(_, drone)| {
+                drone.state() == Some(agent::DroneState::Ready)
+                    && drone.meta.is_some()
+                    && drone.last_seen > Some(min_keepalive)
+            })
+            .map(|(drone_id, _)| drone_id.clone())
+            .collect())
+    }
+
+    pub fn state(&self) -> RwLockReadGuard<WorldState> {
+        self.0.read().expect("Could not acquire world_state lock.")
+    }
+
+    pub fn write_state(&self) -> RwLockWriteGuard<WorldState> {
+        self.0.write().expect("Could not acquire world_state lock.")
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct WorldState {
+    clusters: HashMap<ClusterName, ClusterState>,
+}
+
+impl WorldState {
+    pub fn apply(&mut self, message: WorldStateMessage) {
+        let cluster = self.clusters.entry(message.cluster.clone()).or_default();
+        cluster.apply(message.message);
+    }
+
+    pub fn cluster(&self, cluster: &ClusterName) -> Option<&ClusterState> {
+        self.clusters.get(cluster)
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct ClusterState {
+    pub drones: HashMap<DroneId, DroneState>,
+    pub backends: HashMap<BackendId, BackendState>,
+    pub txt_records: VecDeque<String>,
+}
+
+impl ClusterState {
+    fn apply(&mut self, message: ClusterStateMessage) {
+        match message {
+            ClusterStateMessage::DroneMessage(message) => {
+                let drone = self.drones.entry(message.drone).or_default();
+                drone.apply(message.message);
+            }
+            ClusterStateMessage::BackendMessage(message) => {
+                let backend = self.backends.entry(message.backend).or_default();
+                backend.apply(message.message);
+            }
+            ClusterStateMessage::AcmeMessage(message) => {
+                if self.txt_records.len() > 10 {
```

Comment:
Should this be a `>=`? Or do we intend the txt_records to scale up to 11? (Probably doesn't really matter?)